packages/langium/src/workspace/document-builder.ts

```diff
@@ -104,7 +105,38 @@ export class DefaultDocumentBuilder implements DocumentBuilder {
 
     async build<T extends AstNode>(documents: Array<LangiumDocument<T>>, options: BuildOptions = {}, cancelToken = CancellationToken.None): Promise<void> {
         for (const document of documents) {
-            this.buildState.delete(document.uri.toString());
+            const key = document.uri.toString();
+            if (document.state === DocumentState.Validated) {
+                if (typeof options.validation === 'boolean' && options.validation) {
+                    // Force re-running all validation checks
+                    document.state = DocumentState.IndexedReferences;
+                    document.diagnostics = undefined;
+                    this.buildState.delete(key);
+                } else if (typeof options.validation === 'object') {
+                    const state = this.buildState.get(key);
+                    const previousValidationOptions = state?.options?.validation;
+                    if (typeof previousValidationOptions === 'object' && previousValidationOptions.categories) {
+                        // Validation with explicit options was requested for a document that has already been partly validated.
+                        // In this case, we need to merge the previous validation categories with the new ones.
+                        const newCategories = options.validation.categories ?? ValidationCategory.all;
+                        const previousCategories = previousValidationOptions.categories;
+                        const categories = newCategories.filter(c => !previousCategories.includes(c));
+                        this.buildState.set(key, {
+                            completed: false,
+                            options: {
+                                validation: {
+                                    ...options.validation,
+                                    categories
+                                }
+                            }
+                        });
+                        document.state = DocumentState.IndexedReferences;
```

Comment:
This method of identifying missing categories lacks a real delta. See the following test:

```ts

await builder.build([document1], { validation: { categories: ['built-in', 'fast'] } });
expect(document1.state).toBe(DocumentState.Validated);
expect(document1.diagnostics?.map(d => d.message)).toEqual([
    'Value is too large: 11'
]);

await builder.build([document1], { validation: { categories: ['slow'] } });
expect(document1.state).toBe(DocumentState.Validated);
expect(document1.diagnostics?.map(d => d.message)).toEqual([
    'Value is too large: 11',
    'Bar is too long: AnotherStrangeBar'
]);

await builder.build([document1], { validation: { categories: ['fast'] } });
expect(document1.state).toBe(DocumentState.Validated);
expect(document1.diagnostics?.map(d => d.message)).toEqual([
    'Value is too large: 11',
    'Bar is too long: AnotherStrangeBar'
]);
```

During the last step (running `fast` again), we actually get 3 diagnostics (we get `'Value is too large: 11'` twice), since the document builder "forgot" that we have run `fast` already, and only remembers that we ran `slow`.